simplistic_editor/lib/basic_text_input_client.dart

```diff
@@ -780,6 +792,136 @@ class BasicTextInputClientState extends State<BasicTextInputClient>
+                    : null,
+                _contextMenuAnchors,
+              );
+            },
+      magnifierConfiguration: TextMagnifierConfiguration.disabled,
+    );
+
+    return selectionOverlay;
+  }
+
+  void _toggleToolbar() {
+    final TextSelectionOverlay selectionOverlay =
+        _selectionOverlay ??= _createSelectionOverlay();
+
+    if (selectionOverlay.toolbarIsVisible) {
+      hideToolbar(false);
+    } else {
+      showToolbar();
+    }
+  }
+
+  // When the framework's text editing value changes we should update the text editing
+  // value contained within the selection overlay or we might observe unexpected behavior.
+  void _updateOrDisposeOfSelectionOverlayIfNeeded() {
+    if (_selectionOverlay != null) {
+      if (_hasFocus) {
+        _selectionOverlay!.update(_value);
+      } else {
+        _selectionOverlay!.dispose();
+        _selectionOverlay = null;
+      }
+    }
+  }
+
+  /// Gets the line heights at the start and end of the selection for the given
+  /// editable.
+  _GlyphHeights _getGlyphHeights() {
+    final TextSelection selection = textEditingValue.selection;
+
+    // Only calculate handle rects if the text in the previous frame
+    // is the same as the text in the current frame. This is done because
+    // widget.renderObject contains the renderEditable from the previous frame.
+    // If the text changed between the current and previous frames then
+    // widget.renderObject.getRectForComposingRange might fail. In cases where
+    // the current frame is different from the previous we fall back to
+    // renderObject.preferredLineHeight.
+    final InlineSpan span = renderEditable.text!;
+    final String prevText = span.toPlainText();
+    final String currText = textEditingValue.text;
+    if (prevText != currText || !selection.isValid || selection.isCollapsed) {
+      return _GlyphHeights(
+        start: renderEditable.preferredLineHeight,
+        end: renderEditable.preferredLineHeight,
+      );
+    }
+
+    final String selectedGraphemes = selection.textInside(currText);
+    final int firstSelectedGraphemeExtent =
+        selectedGraphemes.characters.first.length;
+    final Rect? startCharacterRect =
+        renderEditable.getRectForComposingRange(TextRange(
+      start: selection.start,
+      end: selection.start + firstSelectedGraphemeExtent,
+    ));
+    final int lastSelectedGraphemeExtent =
+        selectedGraphemes.characters.last.length;
+    final Rect? endCharacterRect =
+        renderEditable.getRectForComposingRange(TextRange(
+      start: selection.end - lastSelectedGraphemeExtent,
+      end: selection.end,
+    ));
+    return _GlyphHeights(
+      start: startCharacterRect?.height ?? renderEditable.preferredLineHeight,
+      end: endCharacterRect?.height ?? renderEditable.preferredLineHeight,
+    );
+  }
+
+  /// Returns the anchor points for the default context menu.
+  TextSelectionToolbarAnchors get _contextMenuAnchors {
```

Comment:
(Thinking about my context menu APIs, not this PR) I wonder if maybe some of this logic shouldn't be tied up in EditableText, so that anyone building their own editor doesn't have to duplicate it like this...

Reply:
That would be great, maybe it could only be tied to `RenderEditable` and have the code live in `TextSelectionToolbarAnchors` class?